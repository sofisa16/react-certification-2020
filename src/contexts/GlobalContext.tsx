import React, {useState, useEffect, useReducer} from 'react';
import {ThemeProvider} from 'styled-components';
import NoSsr from '@material-ui/core/NoSsr';
import {
  createMuiTheme,
  ThemeProvider as MuiThemeProvider,
  Theme,
} from '@material-ui/core/styles';
import {initialState, reducer, FavoriteVideos, Action} from './../components/FavoritesView/addToFavorites';
import {AUTH_STORAGE_KEY, AUTH_AVATAR, FAVORITES_VIDEOS} from './../utils/constants';
import {storage} from './../utils/storage';
import { useAuth0 } from "@auth0/auth0-react";

interface GlobalContextValues {
  search: string;
  setSearch: React.Dispatch<React.SetStateAction<string>>;
  darkState: boolean;
  setDarkState: React.Dispatch<React.SetStateAction<boolean>>;
  theme: Theme;
  favoriteVideos: FavoriteVideos;
  dispatchFav: React.Dispatch<Action>;
  authenticated: boolean;
  setAuthenticated: React.Dispatch<React.SetStateAction<boolean>>;
  avatar: string;
  setAvatar: React.Dispatch<React.SetStateAction<string>>;
}

interface GlobalContextProviderProps {
  children: React.ReactElement | React.ReactElement[];
}

const GlobalContext = React.createContext<GlobalContextValues>({
  search: '',
  setSearch: () => { return; },
  darkState: false,
  setDarkState: () => { return; },
  theme: createMuiTheme({
    palette: {
      type: 'light',
    }
  }),
  favoriteVideos: {},
  dispatchFav: () => { return; },
  authenticated: false,
  setAuthenticated: () => { return; },
  avatar: '',
  setAvatar: () => { return; },
});

const GlobalContextProvider: React.FunctionComponent<GlobalContextProviderProps> =
  (props: GlobalContextProviderProps) => {
    const [search, setSearch] = useState<string>('wizeline');
    const [darkState, setDarkState] = useState<boolean>(false);
    const [theme, setTheme] = useState<Theme>(
      createMuiTheme({
        palette: {
          type: 'light',
        },
      })
    );
    const [favoriteVideos, dispatchFav] = useReducer(reducer, initialState);
    const [authenticated, setAuthenticated] = useState<boolean>(false);
    const [avatar, setAvatar] = useState<string>('');
    const { user, isAuthenticated } = useAuth0();

    useEffect(
      () => {
        setAuthenticated(isAuthenticated);
      },
      [isAuthenticated]
    );

    useEffect(
      () => {
        if (user) {
          setAvatar(user.picture);
        }
      },
      [user]
    );

    useEffect(
      () => {
        setAuthenticated(Boolean(storage.get(AUTH_STORAGE_KEY)));
        setAvatar(String(storage.get(AUTH_AVATAR)));
        dispatchFav({
          type: 'addAll',
          all: JSON.parse(String(storage.get(FAVORITES_VIDEOS))),
        });
      },
      []
    );

    useEffect(
      () => {
        const palletType = darkState ? 'dark' : 'light';
        setTheme(
          createMuiTheme({
            palette: {
              type: palletType,
            }
          })
        );
      },
      [darkState]
    );

    useEffect(
      () => {
        storage.set(FAVORITES_VIDEOS, JSON.stringify(favoriteVideos));
      },
      [favoriteVideos]
    );

    const value: GlobalContextValues = {
      search,
      setSearch,
      darkState,
      setDarkState,
      theme,
      favoriteVideos,
      dispatchFav,
      authenticated,
      setAuthenticated,
      avatar,
      setAvatar,
    };

    return (
      <NoSsr>
        <MuiThemeProvider theme={theme}>
          <ThemeProvider theme={theme}>
            <GlobalContext.Provider
              value={value}
            >
              {props.children}
            </GlobalContext.Provider>
          </ThemeProvider>
        </MuiThemeProvider>
      </NoSsr>
    );
  };

export {GlobalContextProvider, GlobalContext};