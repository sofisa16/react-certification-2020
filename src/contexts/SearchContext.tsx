import React, {useState} from 'react';

interface SearchContextValues {
  search: string;
  setSearch: React.Dispatch<React.SetStateAction<string>>;
}

interface SearchContextProviderProps {
  children: React.ReactElement | React.ReactElement[];
}

const SearchContext = React.createContext<SearchContextValues>({
  search: "",
  setSearch: () => { return; },
});

const SearchContextProvider: React.FunctionComponent<SearchContextProviderProps> =
  (props: SearchContextProviderProps) => {
    const [search, setSearch] = useState<string>('wizeline');
    const value: SearchContextValues = {
      search,
      setSearch,
    };

    return (
      <SearchContext.Provider
        value={value}
      >
        {props.children}
      </SearchContext.Provider>
    );
  };

export {SearchContextProvider, SearchContext};