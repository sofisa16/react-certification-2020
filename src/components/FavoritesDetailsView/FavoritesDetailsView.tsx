import React, {useState, useEffect, useContext} from 'react';
import {useParams} from 'react-router-dom';
import {YouTubeResponseItems} from './../../data-types/YoutubeAPI';
import DetailsView, {DetailsViewParams} from './../DetailsView/DetailsView';
import {GlobalContext} from './../../contexts/GlobalContext';
import NotFound from './../NotFound/NotFound';

function FavoritesDetailsView(): JSX.Element {
  const {videoId} = useParams<DetailsViewParams>();
  const [relatedItems, setRelatedItems] = useState<YouTubeResponseItems[]>([]);
  const {favoriteVideos} = useContext(GlobalContext);
  const [inFavorites, setInFavorites] = useState<boolean>(false);

  useEffect(
    () => {
      if (favoriteVideos[videoId] && Object.keys(favoriteVideos[videoId]).length > 0) {
        const temp = [];
        for(const key in favoriteVideos) {
          if (key !== videoId) {
            temp.push(favoriteVideos[key]);
          }
        }
        setRelatedItems(temp);
        setInFavorites(true);
      }
      else {
        setInFavorites(false);
      }
    },
    [favoriteVideos, videoId]
  );

  return (
    <>
      {
        inFavorites
          ? <DetailsView relatedItems={relatedItems} path='favorites/' data-testid='FavoritesDetailsView' />
          : <NotFound />
      }
    </>
  );
}

export default FavoritesDetailsView;