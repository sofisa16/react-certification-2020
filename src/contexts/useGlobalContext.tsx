import {useContext} from 'react';
import {GlobalContext, GlobalContextValues} from './GlobalContext';

function useGlobalContext(): GlobalContextValues {
  const context = useContext(GlobalContext);

  if (!context) {
    throw new Error('Please use the "useContext" hook with a GlobalContextProvider.')
  }

  return context;
}

export default useGlobalContext;