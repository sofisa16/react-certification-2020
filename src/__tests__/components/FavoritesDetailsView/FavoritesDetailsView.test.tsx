import React from 'react';
import { render } from '@testing-library/react';
import FavoritesDetailsView from '../../../components/FavoritesDetailsView/FavoritesDetailsView';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import {GlobalContextProvider} from '../../../contexts/GlobalContext';
import CssBaseline from '@material-ui/core/CssBaseline';

it('renders FavoritesDetailsView component', () => {
  const history = createMemoryHistory();
  const route = '/';
  history.push(route);

  const tree = render(
    <GlobalContextProvider>
      <CssBaseline>
        <Router history={history}>
          <FavoritesDetailsView />
        </Router>
      </CssBaseline>
    </GlobalContextProvider>
  );
  expect(tree).toMatchSnapshot();
});