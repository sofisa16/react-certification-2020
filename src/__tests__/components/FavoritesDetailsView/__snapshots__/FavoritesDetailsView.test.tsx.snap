// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FavoritesDetailsView component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bqyKva fbzoJc"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
        >
          404
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          Looks like you have followed a broken link or entered a URL that does not exist on this site.
        </p>
        <a
          href="/"
        >
          Back to home
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bqyKva fbzoJc"
    >
      <h2
        class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
      >
        404
      </h2>
      <p
        class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
      >
        Looks like you have followed a broken link or entered a URL that does not exist on this site.
      </p>
      <a
        href="/"
      >
        Back to home
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
